### PR TITLE
refactor: Massive refactor of the plugin.

### LINKED ISSUE #167: Content plays before ad
When content is played, we have a few milliseconds of the ad play first, before the content.

As much as this should be expected when the prerollTimout has been reached, [we already have our timeout set to 1000ms](https://github.com/guardian/frontend/blob/master/static/src/javascripts/bootstraps/enhanced/media/main.js#L254) (overkill, but I wanted to see if it fixed the problem).

[Content with this issue can be seen here](http://www.theguardian.com/commentisfree/video/2016/feb/03/social-mobility-is-cruel-heres-why-video).

I have dug into it and it seems a script pulled in by the `imasdk` pulls in a [Youtube script](https://s.ytimg.com/yts/jsbin/www-embed-player-vfliX6EpK/www-embed-player.js) that isn't ready for some time after page load. That's the only lead I have, although it might be a red herring.

### LINKED ISSUE #449: No changelog
Releases don't have a changelog, neither there is one listing the versions and the changes made.
So currently the only way to see what the changes are between 2 versions is to look at all the commits in between them.

### PR DESCRIPTION
**Pull request by shawnbuso**
Splits the plugin into multiple files.
Now uses reasonable developer and release workflows (outlined in DEVELOPING.md).

Resolves #167.
Resolves #449.
Potentially resolves other issues as well. I'm going to re-test all FRs and bugs after this is submitted to determine what we have fixed and what still needs fixing.

Reviewers: Do me a favor and sanity check this by following the new developer steps in DEVELOPING.md.

### CODE REVIEW

File: examples/playlist/ads.js [hunk 1]
```diff
@@ -123,21 +121,18 @@ Ads.prototype.onAdEvent = function(event) {
 };
 
 Ads.prototype.onPlaylistItemClick = function(event) {
-  if (!this.initialized) {
-    this.init();
-  }
   if (!this.linearAdPlaying) {
+    if (!this.initialized) {
+      // Handles the case where the user loads the page and clicks a playlist item
+      // immediately, never clicks the play button on the player.
```
Comment (ypavlotsky) [inline examples/playlist/ads.js]: nit: ...loads the page, clicks a playlist item immediately and never clicks...

File: examples/playlist/ads.js [hunk 2]
```diff
@@ -123,21 +121,18 @@ Ads.prototype.onAdEvent = function(event) {
...
+      this.init();
+    }
     this.player.ima.setContentWithAdTag(
         this.contents[event.target.id],
         null,
-        true);
```
Comment (ypavlotsky) [inline examples/playlist/ads.js]: is this change intentional?

File: examples/playlist/ads.js [hunk 3]
```diff
@@ -35,37 +34,28 @@ var Ads = function() {
...
   this.options = {
     id: 'content_video',
     adTagUrl: 'http://pubads.g.doubleclick.net/gampad/ads?sz=640x480&' +
         'iu=/124319096/external/ad_rule_samples&ciu_szs=300x250&ad_rule=1&' +
         'impl=s&gdfp_req=1&env=vp&output=xml_vmap1&unviewed_position_start=1&' +
         'cust_params=sample_ar%3Dpremidpostpod%26deployment%3Dgmf-js&' +
-        'cmsid=496&vid=short_onecue&correlator='
+        'cmsid=496&vid=short_onecue&correlator=',
+    adsManagerLoadedCallback: this.adsManagerLoadedCallback.bind(this),
+    debug: true
```
Comment (ypavlotsky) [inline examples/playlist/ads.js]: We want to leave this on by default?

File: test/index.html [hunk 4]
```diff
@@ -9,11 +9,11 @@
     <link rel="stylesheet" href="//code.jquery.com/qunit/qunit-1.14.0.css">
     <script src="//code.jquery.com/qunit/qunit-1.14.0.js"></script>
     <!-- Load videojs-contrib-ads -->
-    <script src="../node_modules/videojs-contrib-ads/dist/videojs.ads.js"></script>
+    <script src="../node_modules/videojs-contrib-ads/dist/videojs-contrib-ads.js"></script>
     <!-- Load IMA library -->
     <script src="//s0.2mdn.net/instream/html5/ima3_debug.js"></script>
     <!-- Load local IMA plugin -->
-    <link rel="stylesheet" href="../src/videojs.ima.css">
+    <link rel="stylesheet" href="../src/videojs.ima.dev.css">
```
Comment (ypavlotsky) [inline test/index.html]: is this intentional?

File: .eslintrc.js [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+        'sourceType': 'module',
+    },
+    'rules': {
+        'jsdoc/check-types': 'error',
+        'space-infix-ops': 'error',
+    },
+    'plugins': [
+      'jsdoc',
+    ],
+};
```
[the remark this answers is not in the document]
Comment (shawnbuso) [reply .eslintrc.js]: Fixed.

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -16,6 +16,12 @@ Follow either of the two links above to access the appropriate CLA and instructi
 - Sign a Contributor License Agreement (see above).
 - Join the [Google Media Framework discussion group](http://groups.google.com/d/forum/google-media-framework).
 - Fork the library, make the changes and send a [pull request](https://help.github.com/articles/using-pull-requests).
+  - Before submitting your pull requests, please revert any changes to the following files:
+    ```
+    dist/videojs.ima.js
```
[the remark this answers is not in the document]
Comment (shawnbuso) [reply CONTRIBUTING.md]: Good catch, this was from before I had dist/ in the gitignore. Fixed.

File: examples/playlist/ads.js (hunk 3 above)
Comment (shawnbuso) [reply examples/playlist/ads.js]: Nope, forgot to change it back. Fixed.

File: examples/playlist/ads.js (hunk 1 above)
Comment (shawnbuso) [reply examples/playlist/ads.js]: Fixed.

File: examples/playlist/ads.js (hunk 2 above)
Comment (shawnbuso) [reply examples/playlist/ads.js]: Yea we now call player.play() in the adsManagerLoaded callback above, so we don't need to content to play on load.

File: test/index.html (hunk 4 above)
Comment (shawnbuso) [reply test/index.html]: It was, but this made me realize I don't have to do things this way anymore, so it's been reverted.

File: DEVELOPING.md [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+## Releasing
+Releases should only be created by the repository owners. If you think we're in
+need of a new release, please open an issue in our [issue
+tracker](https://github.com/googleads/videojs-ima/issues).
+
+To create a new release, run the following:
+
+```
+npm version [major|minor|patch]
+git push
```
Comment (shawnbuso) [inline DEVELOPING.md]: Fix this - incomplete.

File: package.json [hunk 8]
```diff
@@ -1,29 +1,57 @@
...
-    "videojs-contrib-ads": "~5.0.3"
+    "video.js": ">=5.19.2",
+    "videojs-contrib-ads": "~5.1.0"
   },
   "devDependencies": {
-    "grunt": "^0.4.5",
-    "videojs-doc-generator": "0.0.1"
-  },
-  "peerDependencies": {
-    "video.js": ">=5.3.0"
```
Comment (patrick-mcdougle) [inline package.json]: This is a regression of #417 is there a reason that video.js can't be a peer dependency anymore? Seems weird that a plugin for a thing should also include the thing. Thoughts?
